Add recipe for elcute

viliaapro opened the pull request:
### Brief summary of what the package does

Slurpbarf and Elcute are Emacs minor modes together providing a Paredit-like experience in Lisp Data and nXML modes.

Elcute remaps `kill-line` to a killing command respecting expression structure.

### Direct link to the package repository

https://codeberg.org/vilij/slurpbarf-elcute

### Your association with the package

I am the maintainer.

### Relevant communications with the upstream package maintainer

[e.g., `package.el` compatibility changes that you have submitted. Write **None needed** if there is no problem.]

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

terlar commented:
Seems a bit in the same realm as: https://github.com/AmaiKinono/puni

Might be worth a comparison.

viliaapro commented:
Yes, I started out with Puni, but it didn't work well with nXML mode. My solution targets specifically Lisp Data and XML, and so the code is short and quite simple with a few special cases here and there.

3. heinäkuuta 2024 16.37.52 GMT+03:00 Terje Larsen ***@***.***> kirjoitti:
>Seems a bit in the same realm as: https://github.com/AmaiKinono/puni
>
>Might be worth a comparison.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/melpa/melpa/pull/9088#issuecomment-2206099725
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

riscy commented:
Thanks for this -

> but it didn't work well with nXML mode

I realize it may not be straightforward, but can you describe how this is the case?  This could help the author out in case there are features that could be added to the original package to improve the user experience.

In the meantime, quick first pass --

`elcute.el` with byte-compile using Emacs 29.4:
```
elcute.el:67:7: Warning: reference to free variable `xmltok-type'
In end of data:
elcute.el:53:32: Warning: the function `cl-constantly' might not be defined at runtime.
```

viliaapro commented:
Thank you very much for your attention and feedback!

> > but it didn't work well with nXML mode  
> I realize it may not be straightforward, but can you describe how this is the case?  This could help the author out in case there are features that could be added to the original package to improve the user experience.

There was the original issue that prompted me to writing these two packages: https://github.com/AmaiKinono/puni/issues/50
It had to do with nXML using symbol syntax class for `<` and `>`.  Eventually it got somewhat fixed, but by that time I had already written my own packages, that do the right thing quite cleanly and simply. I've used them for about a year now, and am confident others might find them useful.

Puni has a lot of magic, and I think Lisp Data and XML, being very regular in structure, are better served by a different architecture and approach.  Lisp Data mode is in a way the quintessential Prog mode, and nXML mode is the quintessential Text mode.  To the extent that other languages can be navigated in a similar regular fashion, we can write Paredit-style commands for them.  C and LaTeX come to mind; to the extent that we had well-behaved navigation commands, perhaps with the aid of Tree-sitter, we could make the behave like Lisp and XML.

As it stands, C mode's sexp-traversing commands handle semicolons as not being part of s-expressions, so Elcute doesn't work in C mode.

> In the meantime, quick first pass --
> `elcute.el` with byte-compile using Emacs 29.4:
> ```
> elcute.el:67:7: Warning: reference to free variable `xmltok-type'
> elcute.el:53:32: Warning: the function `cl-constantly' might not be defined at runtime.
> ```

Why don't I get these warnings? I use the command `byte-compile-file`. They warnings might disappear with `(require 'nxml-mode)` and `(require 'cl-lib)`, though.

riscy commented:
To be fair to the author, that issue says "fixed" and you say the issue is "somewhat fixed".  I'm not disputing someone might find your mode useful given the different implementation, of course, but I really do try to validate situations like these to some reasonable degree.

> Why don't I get these warnings?

Most likely you're not seeing all of the compilation warnings because you're compiling your code in an instance of Emacs that has already loaded the code (or its dependencies), or some past version of it at least.  One suggestion is to use `emacs -Q` and try compiling from there.

In your case those `require`s are probably exactly what you need.  I think there have been some changes since the last time I ran the byte-compile, so here's the latest -

`elcute.el` with byte-compile using Emacs 29.4:
```
elcute.el:68:9: Warning: reference to free variable `xmltok-type'
In elcute--nxml-context:
elcute.el:86:20: Warning: reference to free variable `xmltok-type'
In end of data:
elcute.el:85:12: Warning: the function `nxml-token-after' is not known to be defined.
elcute.el:54:32: Warning: the function `cl-constantly' might not be defined at runtime.
```

viliaapro commented:
> To be fair to the author, that issue says "fixed" and you say the issue is "somewhat fixed".  I'm not disputing someone might find your mode useful given the different implementation, of course, but I really do try to validate situations like these to some reasonable degree.

It worked semantically speaking, but I recall the result was not indented to my taste.  Whitespace handling in nXML mode took me the most trial and error in Slurpbarf: indenting, unindenting, skipping trailing blanks.  In Elcute, the creeping logic in Lisp Data mode was the trickiest, because comments are handled asymmetrically in Lisp.  One has to move back a newline after skipping a comment forward to end up at the end of line.

I've fixed some edge cases this weekend; the work seems mostly done apart from proper docstings for the modes themselves.  Any guidelines for those?

Some areas in Elcute are cauterized with user errors, such as using the commands inside Lisp comments and inside processing instructions in nXML mode.  Might change the behaviors later, but they are not very important.

> > Why don't I get these warnings?  

> One suggestion is to use `emacs -Q` and try compiling from there.
> In your case those `require`s are probably exactly what you need.

Thanks, fixed it!

riscy commented:
Thank you!

> Any guidelines for those?

I'd just encourage you to look at solid existing major modes and their documentation.

> Some areas in Elcute are cauterized with user errors, such as using the commands inside Lisp comments and inside processing instructions in nXML mode. Might change the behaviors later, but they are not very important.

I think these are fine, but you might want to give users more explanation.  E.g. `(user-error "Inside comment")` will just show the user an "Inside comment" message (and stop execution) but you may want to say e.g. "elcute operation halted due to being inside comment")

Not a blocker though - welcome to MELPA. :)

viliaapro commented:
> I think these are fine, but you might want to give users more explanation.  E.g. `(user-error "Inside comment")` will just show the user an "Inside comment" message (and stop execution) but you may want to say e.g. "elcute operation halted due to being inside comment")

I kind of like the subdued tone of Emacs builtins' messages; have to think about it.  Branding with the library name would be too much for me; maybe "Not supported inside comment".

> Not a blocker though - welcome to MELPA. :)

Thank you!  Means a lot.
